benchmarks/benchmark_gpu_sparsity.py

```diff
@@ -0,0 +1,248 @@
+import argparse
+import random
+
+import pandas as pd
+import torch
+import torch.utils.benchmark as benchmark
+import torch.nn.functional as F
+from torch import nn
+from torch.sparse import SparseSemiStructuredTensor, to_sparse_semi_structured
+from tqdm import tqdm
+
+
+torch.set_printoptions(
+    precision=2,
+    threshold=None,
+    edgeitems=16,
+    linewidth=480,
+    profile=None,
+    sci_mode=False,
+)
+
+def create_block_sparse_tensor(M, N, blocksize, sparsity, dtype):
+    assert sparsity <= 1.0 and sparsity >= 0.0, \
+        "sparsity should be a value between 0 and 1"
+    A = torch.bernoulli(torch.full((M//blocksize, N//blocksize),
+                        1 - sparsity, dtype=dtype))
+    A = torch.repeat_interleave(A, blocksize, dim=0)
+    A = torch.repeat_interleave(A, blocksize, dim=1)
+    return A.to(dtype).contiguous().cuda()
+
+def create_semi_structured_tensor(
+    r, c, dtype
+):
+    """
+    This function returns a 1:2 sparse matrix of size (r, c).
+    Note that this means this matrix will also be 2:4 and 4:8 sparse as well.
+    """
+
+    choices = [[0, 1], [1, 0]]
+    mask_entries = [random.choice(choices) for i in range(r * c // 2)]
+
+    mask = (
+        torch.tensor(mask_entries, dtype=dtype)
+        .reshape(r, c)
+        .contiguous()
+    ).cuda()
+    sparse_weight = torch.rand(r, c).to(dtype).cuda() * mask
+    return sparse_weight
+
+
+def benchmark_in_us(f, *args, **kwargs):
```

Comment:
There's a couple caveats to this function. I somewhat trust it less than using cuda synchronize and a for loop. Also I'd report the standard deviation as well. If you have 5us but it's +/- 20us something went wrong. blocked autorange is supposed to help with that, but better to verify and print it.

Reply:
Yeah I think blocked autorange is not great - Actually for the benchmarks, a lot of the time it's just running once, likey as @HDCharles highlighted [here](https://github.com/pytorch-labs/ao_benchmarks/blob/7c147e8579d23dacc540f21d460a650035744823/experimental/segment_anything/benchmark_segment_anything.py#L42). 

I think we can use adaptive autorange instead, wrapped in torch.synchronize() to minimze the variability.